### PR TITLE
New rule S3329: Cipher Block Chaining IVs should be unpredictable - for VB.NET 

### LINKED ISSUE #7562: New rule S3329: Cipher Block Chaining IVs should be unpredictable - for VB.NET
Implement S3329 for VB.NET
* Create VB.NET test cases (based on the C# test cases)
* set `ShouldExecute` to always return true
 * modify rule logic if necessary to accommodate VB.NET test cases
 * Update ITs
 * implement ShouldExecute for VB.NET

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Fixes #7562

RSpec PR: https://github.com/SonarSource/rspec/pull/2473

### CODE REVIEW

**Comment by mary-georgiou-sonarsource** [comment]
@zsolt-kolbay-sonarsource I will take over this PR from here to add the rest.
It's Tim that will review this so no harm done.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb [hunk 1]
```diff
@@ -15,6 +15,8 @@ Class InitializationVectorShouldBeRandom
             Dim constantVector = sa.CreateEncryptor(sa.Key, initializationVectorConstant)   ' Noncompliant  {{Use a dynamically-generated, random IV.}}
             '                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+            constantVector = sa.CREATEENCRYPTOR(sa.Key, initializationVectorConstant)       ' Noncompliant
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb:18]: Also, ad tests with different casings for `IV`.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb.bak
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb.bak:1]: Was this committed by accident?

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
Looks good. Just a few more tests.

**Comment by mary-georgiou-sonarsource** [comment]
@Tim-Pohlmann ignore the rspec - @zsolt-kolbay-sonarsource opened the PR and I'm reviewing it now.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/InitializationVectorShouldBeRandomBase.cs [hunk 2]
```diff
@@ -70,7 +70,7 @@ protected override ProgramState PreProcessSimple(SymbolicContext context)
         && property.Instance is { } propertyInstance
         && state.ResolveCaptureAndUnwrapConversion(propertyInstance).TrackedSymbol() is { } propertyInstanceSymbol
         && IsIVProperty(property, propertyInstanceSymbol)
-        && state[assignment.Value]?.HasConstraint(ByteCollectionConstraint.CryptographicallyWeak) is { }
+        && state[assignment.Value]?.HasConstraint(ByteCollectionConstraint.CryptographicallyWeak) is true
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/InitializationVectorShouldBeRandomBase.cs:73]: This change is not tested. We have to add an UT.
`is { }` allows both, `is true` and `is false`, therefore `is false` scenario is not tested. 

To be extra clear: If I change it back to `is { }`, the new UT should fail.

**Review by pavel-mikula-sonarsource — DISMISSED** [review]
We need to test the `is true` change.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.cs [hunk 3]
```diff
@@ -263,14 +263,20 @@ public class ClassWithStaticProperty
...
-
+        aes.IV = customAlg.IV; // Repro for AD0001
         aes.CreateEncryptor(customAlg.Key, customAlg.IV);
+
+        var customIV = new byte[16];
+        var rng = new RNGCryptoServiceProvider();
+        rng.GetBytes(customIV);
+
+        aes.IV = customIV;
+        aes.CreateEncryptor();
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.cs:280]: This here raises an FP if we check that the bool result of `HasConstraint` is just not null and not equal to true.

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.cs:280]: Looks good. I'll dismiss my review as I don't fully revewi this.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+            sa.IV = initializationVectorConstant
+            sa.GenerateKey()
+
+            Dim ivReplacedDefaultConstructor = sa.CreateEncryptor()                         ' FIXME Non-compliant
+            Dim ivReplaced = sa.CreateEncryptor(sa.Key, sa.IV)                              ' FIXME Non-compliant
+        End Using
+    End Sub
+
+    Public Sub RandomIsNotCompliant()
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb:82]: `InstanceFromExpression` test still seems missing. Or is it not applicable to VB?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    End Sub
+    
+    Public Sub TestWithDifferentCase()
+        
+        Dim initializationVectorConstant = New Byte(15) {}
+
+        Using sa As SymmetricAlgorithm = SymmetricAlgorithm.Create("AES")
+            sa.iv = initializationVectorConstant
+            sa.CREATEENCRYPTOR()                        ' Noncompliant
+            sa.CreateEncryptor()                        ' Noncompliant FP
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb:57]: Why FP?

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM, only minor nitpicks.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb (hunk 4 above)
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InitializationVectorShouldBeRandom.vb:81]: Yes - this expression `(If(condition, sa, sa)).IV = initializationVectorConstant`
is not accepted by the compiler and I could not find any other way to do it in a reasonable amount of time.

**Pull request merged**